### PR TITLE
[fix] #4497, Proposal use exceptions in plugin method onUserLogout

### PR DESCRIPTION
**Pull request by jackkum**
Need sort plugins, plg_user_joomla must be last.

``` php
class PlgUserLogout extends JPlugin
{
    public function onUserLogout($user, $options = array())
    {
            throw new Exception("You shall not pass!");
    }
}
```

### CODE REVIEW

**Comment by wilsonge** [comment]
If an exception is thrown we should probably trigger the onUserFailure plugin event and continue. Just like if there's something returned false now.

**Comment by jackkum** [comment]
@wilsonge Is that better?

**Comment by wilsonge** [comment]
Personally yes :)

**Comment by infograf768** [comment]
Any reason to use $ex instead of $e as usually in core?

**Comment by jackkum** [comment]
it's IDE autocomplete, i'll change it.

File: libraries/cms/application/cms.php [hunk 1]
```diff
@@ -926,16 +926,23 @@ public function logout($userid = null, $options = array())
...
+			{
+				$options['username'] = $user->get('username');
+				$this->triggerEvent('onUserAfterLogout', array($options));
+
+				return true;
+			}
+		}
+		catch (Exception $e)
+		{
+			$this->enqueueMessage($e->getMessage());
```
Comment (wilsonge) [inline libraries/cms/application/cms.php]: This needs to return and bail like when you return false as a result

File: libraries/legacy/application/application.php [hunk 2]
```diff
@@ -756,15 +756,23 @@ public function logout($userid = null, $options = array())
...
-				$results = $this->triggerEvent('onUserAfterLogout', array($options));
+				$this->triggerEvent('onUserAfterLogout', array($options));
 
-			return true;
+				return true;
+			}
+		}
+		catch (Exception $e)
+		{
+			$this->enqueueMessage($e->getMessage());
```
Comment (wilsonge) [inline libraries/legacy/application/application.php]: Same thing here as my comment above

**Comment by wilsonge** [comment]
Overall I like this, however I think we need to wrap the `onUserAfterLogout` new plugin event in a separate try/catch. Once a user has passed through `onUserLogout ` then they are logged out and any further exceptions should not mean that `onUserLogoutFailure` will get triggered. Otherwise you're going to get serious bugs

**Comment by jackkum** [comment]
@wilsonge 
if i understand you correctly, i have to do something like this:
```php

// OK, the credentials are built. Lets fire the onLogout event.
try
{
  $results = $this->triggerEvent('onUserLogout', array($parameters, $options));
}
catch (Exception $e)
{
  $this->enqueueMessage($e->getMessage());
  
}

// Check if any of the plugins failed. If none did, success.
try 
{
  if (isset($results) && is_array($results) && !in_array(false, $results, true))
  {
    $options['username'] = $user->get('username');
    $this->triggerEvent('onUserAfterLogout', array($options));

    return true;
  }
}
catch(Exception $e)
{
  $this->enqueueMessage($e->getMessage());
}

// Trigger onUserLoginFailure Event.
$this->triggerEvent('onUserLogoutFailure', array($parameters));
```

**Comment by roland-d** [comment]
@jackkum That code snippet looks good to me.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/5603">issues.joomla.org/joomla-cms/5603</a>.</sub>

**Comment by wilsonge** [comment]
In your first try/catch you need to trigger `onUserLogoutFailure` in the catch and return OR define `$results` as false.

**Comment by jackkum** [comment]
@wilsonge 
If i do trigger `onUserLogoutFailure` in the `catch` i have to do this twise, because old plugins can return `false` instead throws exception. 

`$results` is undefined, so `isset()` will return false, this can be not obvious, and maybe better will define `$results` before try/catch.

**Comment by roland-d** [comment]
@jackkum I guess that is the downside of backwards compatibility, we have to take into account both cases.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/5603">issues.joomla.org/joomla-cms/5603</a>.</sub>

**Comment by roland-d** [comment]
Hello @jackkum 

Thank you for your contribution.

The last comment here was on June 4th. Are you still interested in updating this pull request?
If no reply is received within 4 weeks we will close this issue.

Thanks for understanding! <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/5603">issues.joomla.org/tracker/joomla-cms/5603</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
@roland-d 4 weeks are gone.

**Pull request closed without merging**